File: native-ios/Gugu/Core/Runner.swift
import Foundation

// 구구 점프 — 달리기 미니게임 순수 로직 (lib/runner.ts 이식)
// 좌표계는 웹 SVG viewBox(720×260) 기준을 그대로 쓴다. 렌더러가 비율만 맞춰 그린다.

enum RunnerPhase: String { case ready, running, paused, over }
enum RunnerOutcome: String { case correct, wrong, missed }

struct RunnerQuestion: Equatable {
    var a: Int
    var b: Int
    var choices: [Int]

    var answer: Int { a * b }
}

struct RunnerState {
    var phase: RunnerPhase
    var table: Int?              // nil = 전체 구구단
    var question: RunnerQuestion
    var recentKeys: [String]
    var round: Int
    var score: Int
    var lives: Int
    var combo: Int
    var maxCombo: Int
    var distance: Double
    var obstacleX: Double
    var outcome: RunnerOutcome?
    var given: Int?
    var hitMs: Double
}

enum Runner {
    static let obstacleStart: Double = 690
    static let collisionX: Double = 154
    static let jumpStart: Double = 280
    static let jumpEnd: Double = 30
    /// 피격 연출/정답 공개 유지 시간
    static let hitDurationMs: Double = 1400
    static let maxLives = 3
    /// 장애물 3개를 넘을 때마다 한 단계 빨라진다
    static let scorePerLevel = 3
    /// 정답 뒤 장애물을 넘어가는 속도(px/ms)
    private static let clearSpeed: Double = 0.45

    static func shuffled(_ values: [Int], random: () -> Double = { Double.random(in: 0..<1) }) -> [Int] {
        var result = values
        var i = result.count - 1
        while i > 0 {
            let j = Int(random() * Double(i + 1))
            result.swapAt(i, min(j, i))
            i -= 1
        }
        return result
    }

    /// 정답 1개 + 그럴듯한 오답 2개를 섞은 보기
    static func question(table: Int?, recentKeys: [String], random: () -> Double = { Double.random(in: 0..<1) }) -> RunnerQuestion {
        let problem = Problems.pick(table: table, wrongPool: [:], recentKeys: recentKeys, random: random)
        let answer = problem.a * problem.b
        var seen = Set<Int>()
        let alternatives = [
            answer - problem.a, answer + problem.a, answer - problem.b,
            answer + problem.b, answer - 1, answer + 1,
        ].filter { value in
            guard value > 0, value <= 81, value != answer else { return false }
            return seen.insert(value).inserted
        }
        let picked = Array(shuffled(alternatives, random: random).prefix(2))
        return RunnerQuestion(a: problem.a, b: problem.b, choices: shuffled([answer] + picked, random: random))
    }

    static func create(table: Int?, phase: RunnerPhase = .running, random: () -> Double = { Double.random(in: 0..<1) }) -> RunnerState {
        RunnerState(
            phase: phase,
            table: table,
            // ready 상태의 문제는 화면 예시용 — 시작 시 새로 뽑는다
            question: phase == .ready
                ? RunnerQuestion(a: 2, b: 3, choices: [4, 6, 8])
                : question(table: table, recentKeys: [], random: random),
            recentKeys: [], round: 0, score: 0, lives: maxLives, combo: 0, maxCombo: 0,
            distance: 0, obstacleX: obstacleStart, outcome: nil, given: nil, hitMs: 0
        )
    }

    static func level(score: Int) -> Int { score / scorePerLevel }

    /// 정답 제한 시간 — 한 단계 오를 때마다 500ms 짧아지고 2.8초가 하한
    static func answerWindowMs(score: Int) -> Double {
        max(2800, 5000 - Double(level(score: score)) * 500)
    }

    /// 방금 넘은 장애물로 속도 단계가 올랐는지 (다음 장애물이 막 나온 순간)
    static func leveledUp(_ state: RunnerState) -> Bool {
        state.combo > 0 && state.outcome == nil && state.score > 0 && state.score % scorePerLevel == 0
    }

    static func answer(_ state: RunnerState, choice: Int) -> RunnerState {
        guard state.phase == .running, state.outcome == nil, state.question.choices.contains(choice) else { return state }
        var next = state
        next.given = choice
        next.outcome = choice == state.question.answer ? .correct : .wrong
        return next
    }

    private static func nextObstacle(_ state: RunnerState, random: () -> Double) -> RunnerState {
        var next = state
        next.recentKeys = Array((state.recentKeys + [Problems.key(state.question.a, state.question.b)]).suffix(3))
        next.question = question(table: state.table, recentKeys: next.recentKeys, random: random)
        next.round = state.round + 1
        next.obstacleX = obstacleStart
        next.outcome = nil
        next.given = nil
        next.hitMs = 0
        return next
    }

    /// dt(ms) 만큼 진행. 피격 연출 중이면 연출 시간만 흘리고, 통과하면 점수·콤보를 올린다.
    static func advance(_ state: RunnerState, dt: Double, random: () -> Double = { Double.random(in: 0..<1) }) -> RunnerState {
        guard state.phase == .running, dt.isFinite, dt > 0 else { return state }

        if state.hitMs > 0 {
            let hitMs = max(0, state.hitMs - dt)
            if hitMs > 0 {
                var next = state
                next.hitMs = hitMs
                return next
            }
            if state.lives == 0 {
                var next = state
                next.hitMs = 0
                next.phase = .over
                return next
            }
            return nextObstacle(state, random: random)
        }

        // 정답을 고르기 전에는 제한 시간에 맞춘 속도, 정답 후에는 느리게 흘려보낸다
        let speed = state.outcome == nil
            ? (obstacleStart - collisionX) / answerWindowMs(score: state.score)
            : clearSpeed
        let movement = speed * dt
        var next = state
        next.obstacleX = state.obstacleX - movement
        next.distance = state.distance + movement / 20

        if next.obstacleX <= collisionX && next.outcome != .correct {
            next.obstacleX = collisionX
            next.outcome = next.outcome ?? .missed
            next.lives = next.lives - 1
            next.combo = 0
            next.hitMs = hitDurationMs
            return next
        }
        if next.obstacleX < -40 {
            next.score = next.score + 1
            next.combo = next.combo + 1
            next.maxCombo = max(next.maxCombo, next.combo)
            return nextObstacle(next, random: random)
        }
        return next
    }

    /// 장애물 위치에 점프 높이를 연결해 속도가 올라가도 같은 궤적으로 넘는다
    static func jump(_ state: RunnerState) -> Double {
        guard state.outcome == .correct, state.obstacleX <= jumpStart, state.obstacleX >= jumpEnd else { return 0 }
        return sin(Double.pi * (jumpStart - state.obstacleX) / (jumpStart - jumpEnd)) * 96
    }

    /// 남은 시간 비율 (1 = 방금 등장, 0 = 충돌 지점)
    static func remaining(_ state: RunnerState) -> Double {
        min(1, max(0, (state.obstacleX - collisionX) / (obstacleStart - collisionX)))
    }
}

File: native-android/app/src/main/java/site/smap/gugudan/core/Runner.kt
package site.smap.gugudan.core

import kotlin.math.PI
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.random.Random

// 구구 점프 — 달리기 미니게임 순수 로직 (lib/runner.ts / Core/Runner.swift 이식)
// 좌표계는 웹 SVG viewBox(720×260) 기준을 그대로 쓴다. 렌더러가 비율만 맞춰 그린다.

enum class RunnerPhase { READY, RUNNING, PAUSED, OVER }
enum class RunnerOutcome { CORRECT, WRONG, MISSED }

data class RunnerQuestion(val a: Int, val b: Int, val choices: List<Int>) {
    val answer: Int get() = a * b
}

data class RunnerState(
    val phase: RunnerPhase,
    val table: Int?,              // null = 전체 구구단
    val question: RunnerQuestion,
    val recentKeys: List<String> = emptyList(),
    val round: Int = 0,
    val score: Int = 0,
    val lives: Int = Runner.MAX_LIVES,
    val combo: Int = 0,
    val maxCombo: Int = 0,
    val distance: Double = 0.0,
    val obstacleX: Double = Runner.OBSTACLE_START,
    val outcome: RunnerOutcome? = null,
    val given: Int? = null,
    val hitMs: Double = 0.0,
)

object Runner {
    const val OBSTACLE_START = 690.0
    const val COLLISION_X = 154.0
    const val JUMP_START = 280.0
    const val JUMP_END = 30.0

    /** 피격 연출/정답 공개 유지 시간 */
    const val HIT_DURATION_MS = 1400.0
    const val MAX_LIVES = 3

    /** 장애물 3개를 넘을 때마다 한 단계 빨라진다 */
    const val SCORE_PER_LEVEL = 3

    /** 정답 뒤 장애물을 넘어가는 속도(px/ms) */
    private const val CLEAR_SPEED = 0.45

    fun shuffled(values: List<Int>, random: () -> Double = { Random.nextDouble() }): List<Int> {
        val result = values.toMutableList()
        var i = result.size - 1
        while (i > 0) {
            val j = min((random() * (i + 1)).toInt(), i)
            val tmp = result[i]
            result[i] = result[j]
            result[j] = tmp
            i -= 1
        }
        return result
    }

    /** 정답 1개 + 그럴듯한 오답 2개를 섞은 보기 */
    fun question(
        table: Int?,
        recentKeys: List<String>,
        random: () -> Double = { Random.nextDouble() },
    ): RunnerQuestion {
        val problem = Problems.pick(table, emptyMap(), recentKeys, random)
        val answer = problem.a * problem.b
        val alternatives = listOf(
            answer - problem.a, answer + problem.a, answer - problem.b,
            answer + problem.b, answer - 1, answer + 1,
        ).distinct().filter { it in 1..81 && it != answer }
        val picked = shuffled(alternatives, random).take(2)
        return RunnerQuestion(problem.a, problem.b, shuffled(listOf(answer) + picked, random))
    }

    fun create(
        table: Int?,
        phase: RunnerPhase = RunnerPhase.RUNNING,
        random: () -> Double = { Random.nextDouble() },
    ): RunnerState = RunnerState(
        phase = phase,
        table = table,
        // READY 상태의 문제는 화면 예시용 — 시작 시 새로 뽑는다
        question = if (phase == RunnerPhase.READY) RunnerQuestion(2, 3, listOf(4, 6, 8))
        else question(table, emptyList(), random),
    )

    fun level(score: Int): Int = score / SCORE_PER_LEVEL

    /** 정답 제한 시간 — 한 단계 오를 때마다 500ms 짧아지고 2.8초가 하한 */
    fun answerWindowMs(score: Int): Double = max(2800.0, 5000.0 - level(score) * 500.0)

    /** 방금 넘은 장애물로 속도 단계가 올랐는지 (다음 장애물이 막 나온 순간) */
    fun leveledUp(state: RunnerState): Boolean =
        state.combo > 0 && state.outcome == null && state.score > 0 && state.score % SCORE_PER_LEVEL == 0

    fun answer(state: RunnerState, choice: Int): RunnerState {
        if (state.phase != RunnerPhase.RUNNING || state.outcome != null || choice !in state.question.choices) return state
        return state.copy(
            given = choice,
            outcome = if (choice == state.question.answer) RunnerOutcome.CORRECT else RunnerOutcome.WRONG,
        )
    }

    private fun nextObstacle(state: RunnerState, random: () -> Double): RunnerState {
        val recentKeys = (state.recentKeys + Problems.key(state.question.a, state.question.b)).takeLast(3)
        return state.copy(
            recentKeys = recentKeys,
            question = question(state.table, recentKeys, random),
            round = state.round + 1,
            obstacleX = OBSTACLE_START,
            outcome = null,
            given = null,
            hitMs = 0.0,
        )
    }

    /** dt(ms) 만큼 진행. 피격 연출 중이면 연출 시간만 흘리고, 통과하면 점수·콤보를 올린다. */
    fun advance(state: RunnerState, dt: Double, random: () -> Double = { Random.nextDouble() }): RunnerState {
        if (state.phase != RunnerPhase.RUNNING || !dt.isFinite() || dt <= 0) return state

        if (state.hitMs > 0) {
            val hitMs = max(0.0, state.hitMs - dt)
            if (hitMs > 0) return state.copy(hitMs = hitMs)
            return if (state.lives == 0) state.copy(hitMs = 0.0, phase = RunnerPhase.OVER)
            else nextObstacle(state, random)
        }

        // 정답을 고르기 전에는 제한 시간에 맞춘 속도, 정답 후에는 느리게 흘려보낸다
        val speed = if (state.outcome == null) (OBSTACLE_START - COLLISION_X) / answerWindowMs(state.score) else CLEAR_SPEED
        val movement = speed * dt
        val next = state.copy(
            obstacleX = state.obstacleX - movement,
            distance = state.distance + movement / 20,
        )

        if (next.obstacleX <= COLLISION_X && next.outcome != RunnerOutcome.CORRECT) {
            return next.copy(
                obstacleX = COLLISION_X,
                outcome = next.outcome ?: RunnerOutcome.MISSED,
                lives = next.lives - 1,
                combo = 0,
                hitMs = HIT_DURATION_MS,
            )
        }
        if (next.obstacleX < -40) {
            val combo = next.combo + 1
            return nextObstacle(
                next.copy(score = next.score + 1, combo = combo, maxCombo = max(next.maxCombo, combo)),
                random,
            )
        }
        return next
    }

    /** 장애물 위치에 점프 높이를 연결해 속도가 올라가도 같은 궤적으로 넘는다 */
    fun jump(state: RunnerState): Double {
        if (state.outcome != RunnerOutcome.CORRECT || state.obstacleX > JUMP_START || state.obstacleX < JUMP_END) return 0.0
        return sin(PI * (JUMP_START - state.obstacleX) / (JUMP_START - JUMP_END)) * 96
    }

    /** 남은 시간 비율 (1 = 방금 등장, 0 = 충돌 지점) */
    fun remaining(state: RunnerState): Double =
        min(1.0, max(0.0, (state.obstacleX - COLLISION_X) / (OBSTACLE_START - COLLISION_X)))
}
